std/string_utils.zig

```diff
@@ -0,0 +1,230 @@
+    return strip(a, " \t\n\r", sides);
+}
+
+fn impl_strip_side(a: []const u8, characters: []const u8, start: usize, change: usize, decrement: bool) usize {
+    var moved = true;
+    var index = start;
+
+    while (moved) {
+        moved = false;
+        for (characters) |char| {
+            if (char == a[index]) {
+                moved = true;
+                if (decrement) index -= change else index += change;
+                break;
+            }
+        }
+    }
+    return index;
+}
+
+// If max is 0 then it'll do forever
+pub fn split(a: []const u8, sep: u8, out: &[][]const u8)void {
+    var actualCount: usize = 0;
+    var previousIndex: usize = 0;
+
+    for (a) |char, i| {
+        if (char == sep) {
+            if (i - previousIndex == 0) {
+                (*out)[actualCount] = "";
+            } else {
+                (*out)[actualCount] = a[previousIndex..i];
+            }
+
+            previousIndex = i + 1;
+            actualCount += 1;
+
+            if (actualCount == out.len) break;
+        }
+    }
+
+    (*out)[actualCount] = a[previousIndex..];
+    actualCount += 1;
+    *out = (*out)[0..actualCount];
+}
+
+// Note: characters is an array of u8 not a string!
+// So passing in "abc" doesn't strip abc it strips a, b, and c
+pub fn strip(a: []const u8, characters: []const u8, sides: Side)[]const u8 {
+    var start: usize = 0;
+    var end: usize = a.len - 1;
+    if (sides == Side.LEFT or sides == Side.BOTH) {
+        // Trim left
+        start = impl_strip_side(a, characters, start, 1, false);
+    }
+
+    if (sides == Side.RIGHT or sides == Side.BOTH) {
+        // Trim right
+        end = impl_strip_side(a, characters, end, 1, true);
+    }
+
+    // +1 to convert to 1-index
+    return a[start..end + 1];
+}
+
+pub fn starts_with(a: []const u8, target: []const u8) bool {
+    // Because we are 0-indexing it not 1-indexing it
+    if (a.len < target.len) return false;
+    var i : usize = 0;
+
+    while (i < target.len) {
+        if (a[i] != target[i]) return false;
+        i += 1;
+    }
+    return true;
+}
+
+pub fn ends_with(a: []const u8, target: []const u8) bool {
+    if (a.len < target.len) return false;
+    var diff : usize = a.len - target.len;
+    var i : usize = a.len - 1;
+
+    while (i >= target.len) {
+        if (a[i] != target[i - diff]) return false;
+        i -= 1;
+    }
+    return true;
+}
+
+pub fn str_eql(a: []const u8, b: []const u8) bool {
+    return mem.eql(u8, a, b);
+}
+
+pub fn is_num(byte: u8) bool {
+    return byte >= '0' and byte <= '9';
+}
+
+pub fn to_upper(byte: u8) u8 {
+    return if(is_ascii_lower(byte)) byte - 32 else byte;
+}
+
+pub fn to_lower(byte: u8) u8 {
+    return if(is_ascii_upper(byte)) byte + 32 else byte;
+}
+
+pub fn is_ascii_letter(byte: u8) bool {
+    return is_ascii_lower(byte) or is_ascii_upper(byte);
+}
+
+pub fn is_ascii_lower(byte: u8) bool {
+    return byte >= 'a' and byte <= 'z';
+}
+
+pub fn is_ascii_upper(byte: u8) bool {
```

Comment:
I'm fairly confident that we should not have asciiUpper and asciiLower in the standard library for these reasons:
 * it's too easy to misuse, when actual strings should be used instead
 * if it really is the odd case where you need ascii upper/ascii lower, writing out the function body is easy enough. it should be enough tedium to get people to consider using the actual string functions.

Do you have an actual use case for ascii upper/lower right now?

Reply:
In the randomizer I use a few [ascii functions](https://github.com/Hejsil/pokemon-randomizer/blob/master/src/ascii.zig)(They take `&const u8` because I use them as fn ptrs to generic functions), because I know the strings I get from the nds headers are ascii. I don't mind not having it in `std` though.

Reply:
They could live in `ascii.zig`. Then you sign up for ascii problems by importing that.

Reply:
I'm ok with that.

Reply:
note that "ascii" is different from `[]u8`. ascii requires the top bit of every byte to be `0`, so if you're doing ascii string processing, you might consider validating that top bit.

those ascii functions you linked are correct, because they check for specific ranges of values, but something like `ascii_to_upper()` should perhaps assert the top bit is 0.

Reply:
One could also just check if it was within 127?  That avoids the much more expensive bit masking operations?  But yes it probably does have to be checked, though I would rather verify data through the locale viewer (which would check just once) rather than an indeterminate amount of times :).